cudax/include/cuda/experimental/__memory_resource/any_resource.h

```diff
@@ -0,0 +1,248 @@
+  {
+    _LIBCUDACXX_ASSERT(this->__static_vtable != nullptr, "copying from a moved-from object");
+    this->__static_vtable->__copy_fn(&this->__object, &__other.__object);
+  }
+
+  basic_any_resource& operator=(basic_any_resource& __other)
+  {
+    return this == &__other ? *this : operator=(basic_any_resource(__other));
+  }
+
+  //! @brief Destroys the stored resource
+  ~basic_any_resource() noexcept
+  {
+    if (this->__static_vtable != nullptr)
+    {
+      this->__static_vtable->__destroy_fn(&this->__object);
+    }
+  }
+
+  void swap(basic_any_resource& __other) noexcept
+  {
+    auto __tmp = _CUDA_VSTD::move(__other);
+    __other    = _CUDA_VSTD::move(*this);
+    *this      = _CUDA_VSTD::move(__tmp);
+  }
+
+  //! @brief Equality comparison between two \c basic_any_resource
+  //! @param __rhs The other \c basic_any_resource
+  //! @return Checks whether both resources have the same equality function stored in their vtable and if so returns
+  //! the result of that equality comparison. Otherwise returns false.
+  _LIBCUDACXX_TEMPLATE(class... _OtherProperties)
+  _LIBCUDACXX_REQUIRES((sizeof...(_Properties) == sizeof...(_OtherProperties))
+                         _LIBCUDACXX_AND __properties_match<_OtherProperties...>)
+  _CCCL_NODISCARD bool operator==(const basic_any_resource<_Alloc_type, _OtherProperties...>& __rhs) const
+  {
+    return (this->__static_vtable->__equal_fn == __rhs.__static_vtable->__equal_fn)
+        && this->__static_vtable->__equal_fn(this->__object, __rhs.__object);
+  }
+
+  //! @brief Inequality comparison between two \c basic_any_resource
+  //! @param __rhs The other \c basic_any_resource
+  //! @return Checks whether both resources have the same equality function stored in their vtable and if so returns
+  //! the inverse result of that equality comparison. Otherwise returns true.
+  _LIBCUDACXX_TEMPLATE(class... _OtherProperties)
+  _LIBCUDACXX_REQUIRES((sizeof...(_Properties) == sizeof...(_OtherProperties))
+                         _LIBCUDACXX_AND __properties_match<_OtherProperties...>)
+  _CCCL_NODISCARD bool operator!=(const basic_any_resource<_Alloc_type, _OtherProperties...>& __rhs) const
+  {
+    return !(*this == __rhs);
+  }
+
+  //! @brief Forwards the stateless properties
+  _LIBCUDACXX_TEMPLATE(class _Property)
+  _LIBCUDACXX_REQUIRES((!property_with_value<_Property>) _LIBCUDACXX_AND _CUDA_VSTD::_One_of<_Property, _Properties...>)
+  friend void get_property(const basic_any_resource&, _Property) noexcept {}
+
+  //! @brief Forwards the stateful properties
+  _LIBCUDACXX_TEMPLATE(class _Property)
+  _LIBCUDACXX_REQUIRES(property_with_value<_Property> _LIBCUDACXX_AND _CUDA_VSTD::_One_of<_Property, _Properties...>)
+  _CCCL_NODISCARD_FRIEND __property_value_t<_Property> get_property(const basic_any_resource& __res, _Property) noexcept
+  {
+    _CUDA_VMR::_Property_vtable<_Property> const& __prop = __res;
+    return __prop.__property_fn(__res.__object);
+  }
+};
```

Comment:
Suggestion: I believe we should add a conversion operator to a `basic_resource_ref`, which would be effectively almost trivial and reduce compile time a lot of we assume that people will actually use this